nemo/collections/nlp/data/language_modeling/megatron/ul2_dataset.py

```diff
@@ -0,0 +1,247 @@
+        sample['text_enc'] = text_enc
+        sample['enc_mask'] = np.concatenate([[1], sample['enc_mask']])
+        return sample
+
+    def build_extreme_masking_training_sample(
+        self, sample, target_seq_length, np_rng,
+    ):
+        """Build training sample.
+        Arguments:
+            sample: A list of sentences in which each sentence is a list token ids.
+            target_seq_length: Desired sequence length.
+            max_seq_length: Maximum length of the sequence. All values are padded to
+                this length.
+            vocab_id_list: List of vocabulary ids. Used to pick a random id.
+            vocab_id_to_token_dict: A dictionary from vocab ids to text tokens.
+            cls_id: Start of example id.
+            sep_id: Separator id.
+            mask_id: Mask token id.
+            pad_id: Padding token id.
+            masked_lm_prob: Probability to mask tokens.
+            np_rng: Random number genenrator. Note that this rng state should be
+                  numpy and not python since python randint is inclusive for
+                  the opper bound whereas the numpy one is exclusive.
+            bos_id: start of decoder example id
+            eos_id: end of generation id
+            sentinel_tokens: unique value to be substituted for every replaced span
+            tokenizer_type: wordpiece (BERT-style) or sentencepiece tokenizer. Used for whole word masking logic.
+            max_ngram_size: maximum size of ngrams to be masked.
+            mean_ngram_size: mean size of ngrams to be masked (only used if geometric_dist=True).
+            geometric_dist: Uses a geometric distribution to sample ngram size.
+            permutation: Permutes the ngrams.
+            whole_word_masking: Always masks entire words instead of individual sub-word tokens.
+            favor_long_ngrams: Favor longer ngrams over shorter ones.
+        """
+        assert target_seq_length <= self.max_seq_length
+
+        # flatten sentences into one list
+        tokens = [token for sentence in sample for token in sentence]
+
+        # Truncate to `target_sequence_length`.
+        max_num_tokens = target_seq_length
+        tokens = tokens[:max_num_tokens]
+
+        # Determine if we have a lot of masking or little masking. There are three cases:
+        # 1. Small masking prob, large spans.
+        # 2. Large masking prob, small spans.
+        # 3. Large masking prob, large spans.
+        masked_lm_prob = self.masked_lm_prob if np_rng.randint(0, 3) == 0 else self.extreme_masked_lm_prob
+        if masked_lm_prob == self.masked_lm_prob:
+            # Large spans.
+            max_ngram_size, mean_ngram_size, min_ngram_size = (
+                self.extreme_max_ngram_size,
+                self.extreme_mean_ngram_size,
+                self.extreme_min_ngram_size,
+            )
+        else:
+            # For large masking prob, decide if short or large spans.
+            if np_rng.randint(0, 2) == 0:
```

Comment:
since you have ` np_rng.randint(0, 3)` to determine 3 cases,  this `np_rng.randint(0, 2)` is redundant?

Reply:
Good point. This section was badly written. Refactored it.